src/sqlfluff/dialects/dialect_tsql.py

```diff
@@ -258,7 +258,17 @@
         "NULL",
         Ref("BooleanLiteralGrammar"),
     ),
-    DatatypeIdentifierSegment=Ref("SingleIdentifierGrammar"),
+    DatatypeIdentifierSegment=SegmentGenerator(
+        # Generate the anti template reserved keywords
+        lambda dialect: RegexParser(
+            r"[A-Z][A-Z0-9_]*|\[[A-Z][A-Z0-9_]*\]",
+            CodeSegment,
+            name="data_type_identifier",
+            type="data_type_identifier",
+            anti_template=r"^(" + r"|".join(dialect.sets("reserved_keywords")) + r")$",
```

Comment:
ANSI only excludes `NOT` rather than all keywords. That sounds familiar but can’t remember why. Should we do the same here for consistency?

Reply:
Good question. I'll update the ansi one and see if it's better.

Reply:
I’m wondering if we should do it the other way around? Some data types might be keywords in some dialects. But tests should show that if it’s the case.

Reply:
Looks like it is causing issues. Is there any reason not to just allow keywords here @alanmcruickshank ? Worked fine for ANSI up until now.

Reply:
let's try that too :)